velox/functions/prestosql/ArrayFunctions.h

```diff
@@ -356,4 +356,93 @@ struct ArrayHasDuplicatesFunction {
   }
 };
 
+// Return the union of two arrays without duplicates. The output order is
+// stable, that the elements in each array maintain their original order, and
+// the elements of the second array follow the first one.
+template <typename TExecCtx, typename T>
+struct ArrayUnionFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(TExecCtx);
+
+  template <typename TOutput, typename TInput, typename U>
+  FOLLY_ALWAYS_INLINE void update(
+      TOutput& result,
+      const TInput& array,
+      folly::F14FastSet<U>& uniqSet,
+      bool& hasNull) {
+    for (const auto& item : array) {
+      if (item.has_value()) {
+        if (uniqSet.insert(item.value()).second) {
+          result.add_item() = item.value();
+        }
+      } else if (!hasNull) {
+        hasNull = true;
+        result.add_null();
+      }
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE
+  void update(
+      out_type<velox::Array<Varchar>>& result,
+      const arg_type<velox::Array<Varchar>>& array,
+      folly::F14FastSet<StringView>& uniqSet,
+      bool& hasNull) {
+    for (const auto& item : array) {
+      if (item.has_value()) {
+        if (uniqSet.insert(item.value()).second) {
+          result.add_item().setNoCopy(item.value());
+        }
+      } else if (!hasNull) {
+        hasNull = true;
+        result.add_null();
+      }
+    }
+  }
+
+  template <typename TOutput, typename TInput, typename U>
+  FOLLY_ALWAYS_INLINE void updateNullFree(
+      TOutput& result,
+      const TInput& array,
+      folly::F14FastSet<U>& uniqSet) {
+    for (const auto& item : array) {
+      if (uniqSet.insert(item).second) {
+        result.add_item() = item;
+      }
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE
+  void updateNullFree(
+      out_type<velox::Array<Varchar>>& result,
+      const null_free_arg_type<velox::Array<Varchar>>& array,
+      folly::F14FastSet<StringView>& uniqSet) {
+    for (const auto& item : array) {
+      if (uniqSet.insert(item).second) {
+        result.add_item().setNoCopy(item);
+      }
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE
+  void call(
+      out_type<velox::Array<T>>& result,
```

Comment:
Hi @duanmeng, array_union() needs to support array inputs of any types, including something like ARRAY(ARRAY(BIGINT)) or even ARRAY(ARRAY(MAP(VARCHAR, ROW({BIGINT, SMALLINT})))). I believe our simple function interface does not support writing to a generic-type output right now, so this function needs to be implemented as a vector function.

Reply:
Got it, I will switch to using the vector function, thanks.